abstractions/dotnet/src/RequestInformation.cs

```diff
@@ -118,5 +121,20 @@ public void SetStreamContent(Stream content)
             Headers.Add(ContentTypeHeader, contentType);
             Content = writer.GetSerializedContent();
         }
+        /// <summary>
+        /// Sets the request content from an instance of <see cref="HttpContent"/>
+        /// </summary>
+        /// <param name="httpContent">The <see cref="HttpContent"/> instance to set as a content of the request.</param>
+        /// <param name="cancellationToken">The (optional) <see cref="CancellationToken"/> to use.</param>
+        public async Task SetContentFromHttpContentAsync(HttpContent httpContent, CancellationToken cancellationToken = default)
```

Comment:
Having a dependency on system.net.http in the abstractions breaks of promise of not having implementations details creep up in the abstractions.
If that'd be kind of ok in dotnet because there's a strong consensus on the http client, it wouldn't work for other languages.
We need to find an alternative, would having the content property be a list of streams, and if it contains a single entry, we keep the current behavior, however if we have multiple entries, we build a multipart request work?

Reply:
@baywet I see your point. That would possibly work. I'm afraid though that we might end up creating multiple steps for the user to setup the RequestInformation object.
The other alternative would be to move this to the HttpClient library as an extenstion of the `RequestInformation` class. That way, it wouldn't really creep into the Abstractions and avoid the system.net.http depedency and still leverage the [`MultipartContent`](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.multipartcontent?view=net-5.0) without having to end up creating extra steps for the user to set the RequestInformation content

Reply:
You can check out the latest update and let me know what you think.

Reply:
From a design perspective we can't rely on an extension method. This only works in dotnet and partially works in the TypeScript/Go eco-systems. And I'd like to avoid having language specific implementations for things like this that are so central to the architecture of the SDKs.
Unless you have a better suggestion, let's go with a List of content approach, if people complain about the complexity, we can always try to add facilitating methods in core, which could be in that case an extension method for dotnet/go.

Reply:
No worries. Marking this PR as draft for now as I work this out.

Reply:
Hey @baywet,

After digging into this a bit deeper, I realize a few things: -

- If we choose to build a multipart request from a list of streams, we might still need the ability for the user to define their own boundary string (this would then creep into the `RequestInformation` class). 
- Each individual stream will also need to have the content type defined inside it as a multipart request can contain different content types for each part and we can't assume all the content types will be the same.

We would therefore need to create an abstraction/way to define the content together with the content-type and boundary information (which might end up being a reimplementation of the `MultipartContent` class) so that we can merge the streams appropriately. We could have the alternative assumption that the individual streams already have the content-types defined inside them and all we would do is insert the boundary parameters we receive. 

Alternatively, we could drop this PR and leave it up to the consumer (as below) to create an instance of MultipartContent as they do in dotnet and then call the `ReadAsStreamAsync()` method which they can simply use to set the content and then set the header. This could also be moved down to the relevant core libraries to build the stream containing everything as needed and then just set the stream and header. (This will have the advantage of keeping the abstractions pure and simple from the implementation details of the building of the multipart request).

```cs
var multipartContent = new MultipartContent("mixed", "customBoundary")
{
    new StringContent("value1"), // string content will have the text/plain content header inside
    JsonContent.Create(new { name = "Peter Pan"} ) // json content will have the application/json content header inside
};
requestInformation.Content = await multipartContent.ReadAsStreamAsync(cancellationToken); // set the stream
requestInformation.Headers.Add("Content-Type", "multipart/mixed"); //set the header and we are done.
```

Reply:
Thanks for your continuous investigation on this.
Before we commit to any decision, it would be interesting to see what API Kiota generates for the same scenario.

The current SDK [accepting a stream](https://github.com/microsoftgraph/msgraph-sdk-dotnet/blob/cb13fed8470d529a5c77d4ceb0a7c6bf9094d72b/src/Microsoft.Graph/Requests/Extensions/OnenoteSectionPagesCollectionRequest.cs#L49) (ignoring the multipart content parameter as it's tied to a specific http client).

If Kiota today generates and API surface that accepts a binary array/stream as the body parameter, I don't see any other choice than letting them build the body.

```CSharp
var someBinaryContentWithAllThePartsAndBoundaries = new byte[];
client.Me.OneNote.Section["id"].Pages.PostAsync(someBinaryContentWithAllThePartsAndBoundaries, x => x.Add("Content-Type", "multipart/form-data; boundary=MyPartBoundary198374");
```

However it is very possible today that the generator doesn't do anything with multipart because of #220 that needs to be handled.

Reply:
Thanks for the extra info @baywet,

For the generated code, the current Sdk does not generate the overload accepting the stream (see [here](https://github.com/microsoftgraph/msgraph-sdk-dotnet/blob/cb13fed8470d529a5c77d4ceb0a7c6bf9094d72b/src/Microsoft.Graph/Generated/requests/OnenoteSectionPagesCollectionRequest.cs#L42)). The file you linked to is actually an extension class that has been handwritten as the OneNote Api is abit "weird" as the posting the stream does not happen to its "content" property as its csdl describes it.

This would also explain why running Kiota also generates the post method with the OnenotePage object rather than a binary object (see [here](https://github.com/andrueastman/KiotaSdk/blob/25ddf06cbd509f96a5178450bce7b59d1e9ec05c/Generated/Me/Onenote/Sections/Pages/PagesRequestBuilder.cs#L64)) as well.

If it makes sense, I can close this for now and we can dig further into this after #220. (If need be, one could always send the multipart request with either of the snippets we provided above)

Reply:
so from a service perspective we have 2 main overloads if I understand things right:
- Json body -> the page only contains text
- multipart -> the page contains text and rich content
But today we're only generating for the json body case, is that correct?

Reply:
I'm not 100% sure but I think there is a possibility Json body route might not possibly work and we can't really know since we can't really infer this from the metadata. The docs say we should just post the rich content and has no mention of the json route and it will return the json representation as a response. 
I believe the extensions exist to allow the user to post a `MultipartContent` or `Stream` but unless the metadata somehow says so, we can't avoid generating the json route.

Reply:
so at this point it would be a metadata issue. Is there any other instance of multipart requests in the API this far?

Reply:
After searching through the docs' repo and the SDK repo, I could only find the OneNote api we discussed. 

https://docs.microsoft.com/en-us/graph/api/section-post-pages?view=graph-rest-1.0